### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by evTyurin**
Complete task 4

### CODE REVIEW

**Review by mrseneca — CHANGES_REQUESTED** [review]
Proper git-flow needed.
We agreed to merge 3.4 task into some branch what will be development branch for future.
But your [Electronic_car_service_administrator](https://github.com/evTyurin/SENLA-java-courses/tree/Electronic_car_service_administrator) branch is empty
You can look up Mark's comment about this topic in our skype chat

**Comment by evTyurin** [comment]
My bad. I moved my commits from [Task 3.4](https://github.com/evTyurin/SENLA-java-courses/tree/TASK_3.4) to [Electronic_car_service_administrator](https://github.com/evTyurin/SENLA-java-courses/tree/Electronic_car_service_administrator) previously but didn't push it. Now I fixed it.

File: src/com/senlainc/warsaw/tyurin/entity/Garage.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package com.senlainc.warsaw.tyurin.entity;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class Garage {
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/entity/Garage.java]: Why exactly you need Entity, Repo and Service for Garage? I thought your FULL application should represent some Garage. But right now looks like you have a bunch of garages, with bunch of mechanics. Lets not do this, and change for Garage representation.

Comment (mrseneca) [reply src/com/senlainc/warsaw/tyurin/entity/Garage.java]: So this entity, repo for it and also service should be deleted.

File: src/com/senlainc/warsaw/tyurin/entity/OrderStatus.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.senlainc.warsaw.tyurin.entity;
+
+public enum OrderStatus {
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/entity/OrderStatus.java]: It's not Entity, so you can move it to different package: util or variable for example

File: src/com/senlainc/warsaw/tyurin/entity/Craftsman.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+public class Craftsman{
+
+    private long id;
+    private String name;
+    private String surname;
+    private Map<LocalDateTime, Boolean> schedule;
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/entity/Craftsman.java]: No, scheduling should not be part of the mechanic. "Schedule" will be an abstraction that is based on orders. In the order, you must have the date of its execution, the id of the mechanic and the id of the place in the garage. And based on this, on the go it is necessary to calculate when and who is free on a certain date. There is no need to store it, especially in the entity mechanic

File: src/com/senlainc/warsaw/tyurin/entity/GaragePlace.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.TreeMap;
+
+public class GaragePlace {
+
+    private long id;
+    private int number;
+    private double space;
+    private Map<LocalDateTime, Boolean> schedule;
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/entity/GaragePlace.java]: Same about schedule as for mechanic

File: src/com/senlainc/warsaw/tyurin/service/GaragePlaceService.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+public class GaragePlaceService implements IGaragePlaceService{
+
+    private static GaragePlaceService INSTANCE;
+    private IGaragePlaceDAO garagePlaceDAO;
+    private ICraftsmanDAO craftsmanDAO;
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/service/GaragePlaceService.java]: Unfortunately, it is bad practice to use someone else's dao in an inappropriate service. If you really need to work with the mechanics in this service, you can do it through the mechanics service, not through the dao.

Comment (mrseneca) [reply src/com/senlainc/warsaw/tyurin/service/GaragePlaceService.java]: But think about pass desired mechanic to method what need it, if it's possible.

File: src/com/senlainc/warsaw/tyurin/Test.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+import java.util.List;
+
+public class Test {
+
+    public static void main(String[] args) {
+        GaragePlaceService garagePlaceService = GaragePlaceService.getInstance();
+        CraftsmanService craftsmanService = CraftsmanService.getInstance();
+        OrderService orderService = OrderService.getInstance();
+
+        Craftsman bob = new Craftsman();
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Test.java]: Mechanics, orders and garage places should be created on Service layer, not here (you can count it as Controller layer).
So pass needed variables to service methods and in there create entities and save them to dao

File: src/com/senlainc/warsaw/tyurin/Test.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        order2.setOrderStatus(OrderStatus.CANCELED);
+
+        Order order3 = new Order();
+        order3.setId(3);
+        order3.setStartDate(LocalDateTime.of(2023, 7, 15, 12, 0));
+        order3.setCompletionDate(LocalDateTime.of(2024, 7, 15, 13, 0));
+        order3.addCraftsman(ted);
+        order3.addCraftsman(bob);
+
+        GaragePlaceDAO garagePlaceDAO = GaragePlaceDAO.getInstance();
```
Comment (mrseneca) [inline src/com/senlainc/warsaw/tyurin/Test.java]: You can not operate with dao on this layer, here you can operate only with Service.
So as i wrote in comment before, create and save entities through service methods

**Review by mrseneca — CHANGES_REQUESTED** [review]
Good job, but you need to fix the architecture to make it easier for you to develop in the future and use best practices

**Review by mrseneca — APPROVED** [review]
That will do work for now. Still some small issues in code and architecture, but wil will fix them later. Good job

**Comment by evTyurin** [comment]
> That will do work for now. Still some small issues in code and architecture, but wil will fix them later. Good job

Thank you :)

**Pull request merged**